### DISCUSSION TITLE
Way to set a fixed number of training runs for reinforcement learning

### DISCUSSION THREAD

**Discussion by AstridHornBrorholt** [UPPAALModelChecker/UPPAAL-Meta#232] [Q&A]
By default when using reinforcement learning to generate a strategy using `minE` or `maxE` queries, UPPAAL automatically chooses a number of traces to train on. On a very simple example, it seems to default to 9000 runs. 

![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/11016262/23db85ab-fba5-468d-a610-1dd1b7b2d032)

However, for experiments it can be beneficial to fix the number of training episodes while changing other variables. How can this be done? I have previously been trying something like these learning parameters to train for 10 traces only:

![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/11016262/deb2a1b7-aabd-4a1d-aec9-f76fa9068e8f)

But this (seemingly) results in twice that amount of episodes being trained for. 

![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/11016262/30172f27-a18d-4ee3-ab65-853bf0fe7f76)

Is the number in the result box correct, and should I put in half as many training episodes as I want in the boxes? Or is the reporting misleading somehow, and it actually trained for the 10 traces I wanted?

**Comment by thorulf4** [comment]
Unfortunately we don't have good documentation on these parameters internally, and I think we need help from @petergjoel to completely understand how your desired results.

**Comment by petergjoel** [comment]
Most of the parameters have a direct equivalent in Fig. 2 of [On Time with Minimal Expected Cost!](https://homes.cs.aau.dk/~kgl/SSFT2015/ATVA.pdf).

For your particular case, you need to limit to one iteration, as per your example.

1. limit the number of iterations to `1`,
2. set `Number of successful runs`, `Maximum number of runs` and `Number of good runs` to the desired number of iterations, and
3. set evaluation runs to `1` (the evaluation is only used to discriminate between iterations).

This assume that your model _ALWAYS_ reaches the trace-termination condition.
If not, then you will have to adjust the first three parameters to accommodate for the conditional learning.

**Comment by AstridHornBrorholt** [comment]
Setting `number of runs to evaluate` to 1 changes the number of runs reported from 20 to 11. The UI doesn't allow setting it to 0, but that hardly matters. This will save a good deal of time when running learning queries in the UI. (I see that when calling `veriftya` through CLI, I use the default 100 evaluation runs which is not so much.)

![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/11016262/b776acd6-9b01-44b4-839a-8beb771c7b45)

![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/11016262/9657e463-c5df-4559-af87-d71c57491a0a)
